internal/configfsm/fsm.go

```diff
@@ -0,0 +1,277 @@
+		}
+	case OutOfDate:
+		err = fsm.Fire(OutOfDateToOutOfDate, fsmHelper)
+		if err != nil {
+			return aState, fmt.Errorf("could not Fire state %s, err: %w", aState, err)
+		}
+	}
+	// Update the current state
+	aState, err = fsm.State(ctx)
+	if err != nil {
+		return aState, fmt.Errorf("could not get state, err: %w", err)
+	}
+	return aState, nil
+}
+
+// InitFSM Initialize the state machine algorithm
+func InitFSM(state string) (fsm *stateless.StateMachine, err error) {
+	fsm = stateless.NewStateMachine(state)
+	fsm.Configure(Start).Permit(StartToMissing, Missing)
+	fsm.Configure(Missing).
+		OnEntry(func(_ context.Context, args ...any) error {
+			fsmHelper := args[0].(FsmHelper)
+			fmt.Println("Entering Missing")
+			fsmHelper.ResetNonCompliantAt()
+			if fsmHelper.ArePoliciesMatched() {
+				return fsm.Fire(MissingToClusterNotReady, fsmHelper)
+			}
+			return nil
+		}).
+		Permit(MissingToClusterNotReady, ClusterNotReady).
+		PermitReentry(MissingToMissing)
+	fsm.Configure(ClusterNotReady).
+		OnEntry(func(_ context.Context, args ...any) error {
+			fsmHelper := args[0].(FsmHelper)
+			fmt.Println("Entering ClusterNotReady")
+			if !fsmHelper.ArePoliciesMatched() {
+				return fsm.Fire(ClusterNotReadyToMissing, fsmHelper)
+			}
+			if fsmHelper.IsClusterReady() ||
+				fsmHelper.IsTimedOut() {
+				return fsm.Fire(ClusterNotReadyToInProgress, fsmHelper)
+			}
+			return nil
+		}).
+		Permit(ClusterNotReadyToMissing, Missing).
+		Permit(ClusterNotReadyToInProgress, InProgress).
+		PermitReentry(ClusterNotReadyToClusterNotReady)
+	fsm.Configure(InProgress).
+		OnEntry(func(_ context.Context, args ...any) error {
+			fsmHelper := args[0].(FsmHelper)
+			fmt.Println("Entering InProgress")
+			if fsmHelper.IsNonCompliantAtZero() ||
+				fsmHelper.IsAllPoliciesCompliant() ||
+				(!fsmHelper.IsNonCompliantPolicyInEnforce() && !fsmHelper.IsAllPoliciesCompliant()) {
+				fsmHelper.SetNonCompliantAtNow()
+			}
+			if !fsmHelper.ArePoliciesMatched() {
+				return fsm.Fire(InProgressToClusterNotReady, fsmHelper)
+			}
+			if fsmHelper.IsTimedOut() {
+				return fsm.Fire(InProgressToTimedOut, fsmHelper)
+			}
+			if !fsmHelper.IsClusterReady() {
+				return fsm.Fire(InProgressToClusterNotReady, fsmHelper)
+			}
+			if !fsmHelper.IsNonCompliantPolicyInEnforce() && !fsmHelper.IsAllPoliciesCompliant() {
+				return fsm.Fire(InProgressToOutOfDate, fsmHelper)
+			}
+			if fsmHelper.IsAllPoliciesCompliant() {
+				return fsm.Fire(InProgressToCompleted, fsmHelper)
+			}
+			return nil
+		}).
+		Permit(InProgressToClusterNotReady, ClusterNotReady).
+		Permit(InProgressToTimedOut, TimedOut).
+		Permit(InProgressToOutOfDate, OutOfDate).
+		Permit(InProgressToCompleted, Completed).
+		PermitReentry(InProgressToInProgress)
+	fsm.Configure(OutOfDate).
+		OnEntry(func(_ context.Context, args ...any) error {
+			fsmHelper := args[0].(FsmHelper)
+			fmt.Println("Entering InProgress")
+			fsmHelper.ResetNonCompliantAt()
+			if !fsmHelper.ArePoliciesMatched() ||
```

Comment:
If we're in `outOfDate` because of an `Inform` `NonCompliant` policy, but then we delete it, it should go to `Completed`, through `InProgress`, but neither one of those conditions would facilitate the transition. Kindly confirm.

Reply:
Fixed. Also added a unit test for it